### PR TITLE
[BEAM-14003] Adds compat for Elasticsearch 8.0.0 and clean unsupported versions and previous frameworks leftovers

### PR DESCRIPTION
**Pull request by egalpin**
Adds compatibility for Elasticsearch 8.x with test suite.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [x] Update `CHANGES.md` with noteworthy changes.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by egalpin** [comment]
R: @echauchot

**Comment by egalpin** [comment]
Java pre commit failures seem to indicate that the testcontainer for 6.4.0 couldn't be fetched or interacted with properly.  I tried locally after purging images and all worked as expected. Going to retry.

**Comment by egalpin** [comment]
Run Java PreCommit

**Review by echauchot — CHANGES_REQUESTED** [review]
Thanks @egalpin for this migration and sorry for the delay. 
It seems to be a very simple migration with no breaking changes, only the type was removed at ES side. 
So only minor comments + ES2 removing. Also please go through [BEAM-13136](https://issues.apache.org/jira/browse/BEAM-13136) and apply the cleanings in a separate commit (with jira id) in the current PR (as it is only cleaning leftovers). Please fix the build and spotless etc.

**Comment by egalpin** [comment]
> It seems to be a very simple migration with no breaking changes, only the type was removed at ES side.

@echauchot That's right, not much was needed to make it compatible 👍

**Comment by egalpin** [comment]
Run Java PreCommit

**Comment by egalpin** [comment]
@echauchot I went through the itemized list on BEAM-13136 (there's 2 commits[1][2] with that issue tag in the commit list of this PR).  I was able to remove all 2.x references as listed on BEAM-13136 and did a few searches for other references without finding other references.  I believe all 2.x-specific items have been removed.

[1] 6e02aed0e507156e93d6c43c41f84b1382d61c37
[2] 9a1877f09581ea0083d2344dd852303e6e7977af

**Comment by egalpin** [comment]
Run Java PreCommit

**Comment by echauchot** [comment]
@egalpin I just started [a thread](https://lists.apache.org/thread/70phb0cfy2knoz24h2zoj1go5nvx0g2w) to remove also ES5 and ES6 support as they are no more supported by elastic. Maybe we will do it in this PR as well.

**Comment by egalpin** [comment]
@echauchot It definitely makes sense to remove support/simplify implementation if 5.x and 6.x are end-of-life products. 

I think it might make sense to make a subsequent PR to do that removal so that changes could be separately reverted if needed, and so that this PR doesn’t have the wait for votes or discussion on unrelated versions in order to support the new 8.x version.

Thoughts?

**Comment by echauchot** [comment]
> @echauchot It definitely makes sense to remove support/simplify implementation if 5.x and 6.x are end-of-life products.
> 
> I think it might make sense to make a subsequent PR to do that removal so that changes could be separately reverted if needed, and so that this PR doesn’t have the wait for votes or discussion on unrelated versions in order to support the new 8.x version.
> 
> Thoughts?

+1

**Comment by egalpin** [comment]
@echauchot Any remaining actions items for this PR in that case?

File: sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java [hunk 1]
```diff
@@ -2551,6 +2554,15 @@ public void closeClient() throws IOException {
     }
   }
 
+  private static void maybeLogVersionDeprecationWarning(int clusterVersion) {
+    if (DEPRECATED_CLUSTER_VERSIONS.contains(clusterVersion)) {
+      LOG.warn(
+          "Support for cluster version {} will be dropped in a future release of the "
```
Comment (echauchot) [inline sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:2560]: sed s/cluster/Elasticsearch cluster/g

File: sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java [hunk 2]
```diff
@@ -203,7 +202,9 @@
 })
 public class ElasticsearchIO {
```
Comment (echauchot) [inline sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:203]: please remove all leftover references to shardPreference, it was an ES2 only code

**Review by echauchot — CHANGES_REQUESTED** [review]
Thanks Evan for the fixes ! Only one clean that you forgot and a minor change on the log message and we're done !

**Comment by echauchot** [comment]
@egalpin community metrics failure is unrelated to your PR.

**Review by echauchot — APPROVED** [review]
LGTM ! thanks
Please squash review commits but keep 2 commits as related to 2 separate tickets (ES8, cleaning ES2).
I'll do the merge.

**Comment by egalpin** [comment]
@echauchot done :)

**Comment by echauchot** [comment]
Run Java PreCommit

**Comment by egalpin** [comment]
Thanks @echauchot , looks to have passed now 👍

**Comment by echauchot** [comment]
Community metrics failure is a timeout on an external server. Merging ! Thanks for your work Evan !

**Comment by egalpin** [comment]
Thanks Etienne!

**Pull request merged**